### ISSUE TITLE
etcd server failed to create watch id

### ISSUE THREAD

**Issue by sidelines** [coreos/jetcd#312]
com.coreos.jetcd.common.exception.EtcdException: etcd server failed to create watch id
    at com.coreos.jetcd.common.exception.EtcdExceptionFactory.newEtcdException(EtcdExceptionFactory.java:30)
    at com.coreos.jetcd.internal.impl.WatchImpl.processCreate(WatchImpl.java:271)
    at com.coreos.jetcd.internal.impl.WatchImpl.processWatchResponse(WatchImpl.java:198)
    at com.coreos.jetcd.internal.impl.WatchImpl.access$100(WatchImpl.java:61)
    at com.coreos.jetcd.internal.impl.WatchImpl$1.onNext(WatchImpl.java:177)
    at com.coreos.jetcd.internal.impl.WatchImpl$1.onNext(WatchImpl.java:174)
    at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:406)
    at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:36)
    at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:36)
    at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:36)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:530)
    at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

I uesd jetcd 0.0.2 and catch this error  , the watchId is -1 could throw this exception, I don't find what kind of scene will appear this error， how to solve this ？

**Comment by lburgazzoli** [comment]
Looks like an issue with etcd istself

**Comment by oliugian** [comment]
hello,I have got the same problem.how to solve this?

**Comment by johnxiaohe** [comment]
same problem when reconnect etcd server

**Comment by lburgazzoli** [comment]
can you provide a reproducer ?

**Comment by johnxiaohe** [comment]
> can you provide a reproducer ?
Step
1. open watch
2. stop etcd
3. jetcd will frequently callback Linstener.onError. Exception Message: io.exception
4. start etcd
5. jetcd callback Linstener.onError. ExceptionMessage:  etcd server failed to create watch id

I try rerun watch operating. But cant watch. Exception Message: io.exception

**Comment by lburgazzoli** [comment]
do you have any time to submit a PR with an integration test ?

**Comment by johnxiaohe** [comment]
> do you have any time to submit a PR with an integration test ?

Recently, the project is relatively busy, and there is no time to do the previous PR. And I think the above process steps are very clear.
It is difficult to do integration testing in this scenario

**Comment by johnxiaohe** [comment]
Application Logs:
```text
2021-11-11 18:00:59.759 ERROR 115483 --- [ult-executor-15] i.a.c.a.watch.PubConfigWatchLinstener    : ==========================【Confcenter】 An Error Occurred In The ETCD Monitoring, And The Reason For The Error: etcd server failed to create watch id==========================
^Cio.etcd.jetcd.common.exception.EtcdException: etcd server failed to create watch id
	at io.etcd.jetcd.common.exception.EtcdExceptionFactory.newEtcdException(EtcdExceptionFactory.java:30)
	at io.etcd.jetcd.WatchImpl$WatcherImpl.onNext(WatchImpl.java:203)
	at io.etcd.jetcd.WatchImpl$WatcherImpl.onNext(WatchImpl.java:99)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:436)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:33)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInternal(ClientCallImpl.java:658)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:643)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
ETCD Logs
```text
2021-11-11 18:01:00.158564 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158569 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158575 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158580 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158585 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158593 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158597 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158605 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158609 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158617 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158621 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158628 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158632 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158638 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158644 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158649 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158656 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158661 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158669 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158674 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158680 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158684 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
2021-11-11 18:01:00.158694 D | etcdserver/api/v3rpc: failed to receive watch request from gRPC stream ("rpc error: code = Canceled desc = context canceled")
```

**Comment by johnxiaohe** [comment]
debug get headers:
```text
header {
  cluster_id: 16682809393371496585
  member_id: 8134730084721976868
  revision: 5041
  raft_term: 79
}
watch_id: -1
created: true
canceled: true
cancel_reason: "rpc error: code = PermissionDenied desc = etcdserver: permission denied"
```

**Comment by johnxiaohe** [comment]
I guess because the ETCD restarted. So the previous Token has expired. But returning here does not return an error that the authentication is invalid. Instead, there is no permission. So will not go through the re-authorization process?

**Comment by lburgazzoli** [comment]
it could be, [this](https://github.com/etcd-io/jetcd/blob/aa057f86003f941a8e685e87df4d4d0980a95dd4/jetcd-core/src/main/java/io/etcd/jetcd/Util.java#L84-L87) method should probably be updated.

However the error from etcd is really mileading

**Comment by johnxiaohe** [comment]
> it could be, [this](https://github.com/etcd-io/jetcd/blob/aa057f86003f941a8e685e87df4d4d0980a95dd4/jetcd-core/src/main/java/io/etcd/jetcd/Util.java#L84-L87) method should probably be updated.
> 
> However the error from etcd is really mileading

Yes. And the error of the proxy node returns CODE

**Comment by johnxiaohe** [comment]
This is indeed the problem. After ETCD restarted, I called a one-yuan request to automatically refresh the token. Then the monitoring succeeded. I hope the refresh token method can be called externally. Finally, can this fix be updated to version 0.5.0?

**Comment by lburgazzoli** [comment]
we can probably include the check for that error in the method I linked

**Comment by johnxiaohe** [comment]
When the connection has been disconnected. Failed to close Watcher. At this time the connection has been disconnected.

```text
Exception while executing runnable io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed@44ee5daf

java.lang.IllegalStateException: Stream is already completed, no further calls are allowed
	at com.google.common.base.Preconditions.checkState(Preconditions.java:511) ~[guava-28.1-android.jar!/:na]
	at io.grpc.stub.ClientCalls$CallToStreamObserverAdapter.onNext(ClientCalls.java:345) ~[grpc-stub-1.27.1.jar!/:1.27.1]
	at io.etcd.jetcd.WatchImpl$WatcherImpl.resume(WatchImpl.java:151) ~[jetcd-core-0.5.0.jar!/:na]
	at io.etcd.jetcd.WatchImpl.watch(WatchImpl.java:81) ~[jetcd-core-0.5.0.jar!/:na]
	at io.amber.confcenter.autoconfig.watch.ConfcenterLinuxWatch.startWatch(ConfcenterLinuxWatch.java:37) ~[amberconfcenter-spring-boot-starter-1.0-SNAPSHOT.jar!/:na]
	at io.amber.confcenter.autoconfig.watch.ConfcenterConfigWatch.watch(ConfcenterConfigWatch.java:53) ~[amberconfcenter-spring-boot-starter-1.0-SNAPSHOT.jar!/:na]
	at io.amber.confcenter.autoconfig.watch.PubConfigWatchLinstener.onCompleted(PubConfigWatchLinstener.java:56) ~[amberconfcenter-spring-boot-starter-1.0-SNAPSHOT.jar!/:na]
	at io.etcd.jetcd.WatchImpl$WatcherImpl.close(WatchImpl.java:173) ~[jetcd-core-0.5.0.jar!/:na]
	at io.amber.confcenter.autoconfig.watch.PubConfigWatchLinstener.onError(PubConfigWatchLinstener.java:49) ~[amberconfcenter-spring-boot-starter-1.0-SNAPSHOT.jar!/:na]
	at io.etcd.jetcd.WatchImpl$WatcherImpl.onError(WatchImpl.java:260) ~[jetcd-core-0.5.0.jar!/:na]
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:449) ~[grpc-stub-1.27.1.jar!/:1.27.1]
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39) ~[grpc-api-1.27.1.jar!/:1.27.1]
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23) ~[grpc-api-1.27.1.jar!/:1.27.1]
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40) ~[grpc-api-1.27.1.jar!/:1.27.1]
	at io.etcd.jetcd.ClientConnectionManager$AuthTokenInterceptor$1$1.onClose(ClientConnectionManager.java:289) ~[jetcd-core-0.5.0.jar!/:na]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[grpc-core-1.27.1.jar!/:1.27.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_291]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_291]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_291]
```

**Comment by johnxiaohe** [comment]
Temporary solution

> Linstener callback

```java
@Override
    public void onError(Throwable throwable) {
        LOGGER.error("========================== An Error Occurred In The ETCD Monitoring, And The Reason For The Error: {}==========================", throwable.getMessage());
        try{
            String value = EtcdOperations.getSingle(prefix);
            LOGGER.error("==================Value getd, Value = {}", value);
           // Must reconnect and then close the monitor. 
            watcher.close();
        }catch (Exception e){
            return;
        }
    }

    @Override
    public void onCompleted() {
        LOGGER.warn("==========================【Confcenter】 End Of Monitoring==========================");
        // OnCompleted will be called after closing
        watch.rewatch();
    }
```

**Comment by johnxiaohe** [comment]
etcd V3.4 and last version has  been slove this problem
https://github.com/etcd-io/etcd/issues/10743

**Comment by johnxiaohe** [comment]
I reviewed the issue again. 
I found that the initial problem wat an issue with **expired tokens.**
In version 0.5.0 of Jetcd，failure to handle errors with expired tokens would result in the error message `etcdserver failed to create watch id ` when `watch = -1` is printed.
In version 0.5.4 or lates version，the error of `permission denied` be specially judged. so return "permission denied" msg
In etcd 3.3 versions, the code like 
![image](https://user-images.githubusercontent.com/37014780/232237567-5e051c9a-0c0b-4cd5-8241-42c08327a2bb.png)
and then have a pr, but can`t rslove the problem.  
https://github.com/etcd-io/etcd/pull/9065/files#diff-ac07bc549aa789169e8afd08a258e6e504503fa6fec5d041329772a386560510
i see the refresh token in 0.5.4 version. so maybe we can rewatch in this error happend
I hope this info helps someone in the back
